File: src/232-implement-queue-using-stacks/solution.rb
class MyQueue
    attr_reader :primary_stack, :secondary_stack
    def initialize()
      @primary_stack = [] 
      @secondary_stack = [] 
    end


=begin
    :type x: Integer
    :rtype: Void
=end
    def push(x)
      if empty
        primary_stack << x
      else
        until primary_stack.empty?
          secondary_stack << primary_stack.pop
        end

        primary_stack << x

        until secondary_stack.empty?
          primary_stack << secondary_stack.pop
        end
      end
    end


=begin
    :rtype: Integer
=end
    def pop()
      primary_stack.
    end


=begin
    :rtype: Integer
=end
    def peek()
      primary_stack.last 
    end


=begin
    :rtype: Boolean
=end
    def empty() 
      (primary_stack + secondary_stack).empty?
    end
end

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue.new()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: src/232-implement-queue-using-stacks/solution.ts
class MyQueue {
    private s1: number[];
    private s2: number[];

    constructor() {
      this.s1 = [];
      this.s2 = [];
    }

    push(x: number): void {
      while (this.s2.length > 0) this.s1.push(this.s2.pop()!);

      this.s1.push(x);

      while (this.s1.length > 0) this.s2.push(this.s1.pop()!);
    }

    pop(): number {
      return this.s2.pop();
    }

    peek(): number {
      return this.s2.at(-1);
    }

    empty(): boolean {
      return this.s2.length === 0;
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
